File: cpp/sfo_11_find_minimum_in_rotated_sorted_array_s2/sfo_11_find_minimum_in_rotated_sorted_array_s2.cpp
/*
* File: sfo_11_find_minimum_in_rotated_sorted_array_s2.cpp
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    int minArray(vector<int>& numbers) {
        int i = 0, j = numbers.size() - 1;
        while (i < j) {
            int m = (i + j) / 2;
            if (numbers[m] > numbers[j]) i = m + 1;
            else if (numbers[m] < numbers[j]) j = m;
            else {
                int x = i;
                for(int k = i + 1; k < j; k++) {
                    if(numbers[k] < numbers[x]) x = k;
                }
                return numbers[x];
            }
        }
        return numbers[i];
    }
};

int main() {
    // ======= Test Case =======
    vector<int> numbers = { 3, 4, 5, 1, 2 };
    // ====== Driver Code ======
    Solution* slt = new Solution();
    int res = slt->minArray(numbers);
    cout << res << endl;
    
    return 0;
}

File: sword_for_offer/codes/python/sfo_11_find_minimum_in_rotated_sorted_array_s2.py
'''
File: sfo_11_find_minimum_in_rotated_sorted_array_s2.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def minArray(self, numbers: List[int]) -> int:
        i, j = 0, len(numbers) - 1
        while i < j:
            m = (i + j) // 2
            if numbers[m] > numbers[j]: i = m + 1
            elif numbers[m] < numbers[j]: j = m
            else: return min(numbers[i:j])
        return numbers[i]

# ======= Test Case =======
numbers = [3, 4, 5, 1, 2]
# ====== Driver Code ======
slt = Solution()
res = slt.minArray(numbers)
print(res)
